aten/src/ATen/core/ATenDispatch.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <c10/core/Backend.h>
+#include <unordered_map>
+#include <c10/util/C++17.h>
+#include <memory>
+
+// This dispatch class serves as a replacement for our previous dispatch
+// mechanism, in which all functions were members of a Type class. A derived
+// class existed for each backend (and Variable), and the vtable was used to
+// dispatch to the correct implementation. This class is to be replaced by
+// the c10 dispatcher when it supports all argument and return types.
+// This implementation opts to store implementations in a table of void*.
+
+namespace at {
+
+// ATenOpTable stores the implementations for each backend, in addition to
+// an implementation for variables.
+class CAFFE2_API ATenOpTable {
+ public:
+  ATenOpTable(std::string schema)
+    : schema_(schema) {}
+
+  template<class FuncType>
+  FuncType* getOp(Backend backend, bool is_variable) const {
+    if (is_variable) {
+      return reinterpret_cast<FuncType*>(getVariableOp());
+    }
+    return reinterpret_cast<FuncType*>(getBaseOp(backend));
+  }
+ private:
+  void registerOp(Backend backend, void* fn) {
+    TORCH_CHECK(function_table_[static_cast<int64_t>(backend)] == nullptr,
```

Comment:
Internal asserts here? (I guess if you're going to expose directly to extensions an internal assert is not appropriate, but for internal use, this isn't a public API right)

Reply:
There's no type checking in the registration API right now; as long as we don't have any type checking I have a pretty strong preference of NOT allowing external parties to poke this API. I don't know what the situation is with XLA and this diff at the moment.